More changes to environments

tas50 opened the pull request:
Explain cookbook version pinning

Remove Attribute Persistence which should only be in the attributes doc

Remove the () from node.environment. This isn't javascript

Move the chef solo instructions to their own category

Make an initial mention of cookbook pinning. This is a super important feature that all our large customers do and we don't even call it out here. Kinda weird

Remove the odd section explaining what Ruby is. We have that earlier on and we don't have a similar section for what JSON is. Eventually we need to get hover overs or aggressive links for this kinda thing. Right now though this just bloats the doc and detracts from the real content.

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
More changes to environments

Explain cookbook version pinning

Remove Attribute Persistence which should only be in the attributes doc

Remove the () from node.environment. This isn't javascript

Move the chef solo instructions to their own category

Make an initial mention of cookbook pinning. This is a super important feature that all our large customers do and we don't even call it out here. Kinda weird

Remove the odd section explaining what Ruby is. We have that earlier on and we don't have a similar section for what JSON is. Eventually we need to get hover overs or aggressive links for this kinda thing. Right now though this just bloats the doc and detracts from the real content.

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Add better version contraint examples

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Cookstyle fixes to the examples

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
Replicate dtag

Signed-off-by: Tim Smith <tsmith@chef.io>

kagarmoe commented on chef_master/source/api_chef_server.rst:
```diff
@@ -2827,7 +2827,7 @@ The response is similar to:
 .. tag environment
 
-An environment is a way to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. Every organization begins with a single environment called the ``_default`` environment, which cannot be modified (or deleted). Additional environments can be created to reflect each organization's patterns and workflow. For example, creating ``production``, ``staging`` ...
+An environment is a way to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. Every organization begins with a single environment called the ``_default`` environment, which cannot be modified (or deleted). Additional environments can be created to reflect each organization's patterns and workflow. For example, creating ``production``, ``staging`` ...
```
An environment is a way to map an organization's real-life workflow TO WHAT can be configured and managed when using Chef Infra.

Can we find a more useful phrasing? TO WHAT is obscure. 

Environments are useful for mapping an organization's real-life workflows to Chef Infra configurations and management definitions.

This is awkward, but a beginning.

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -5,21 +5,21 @@ About Environments
 The _default Environment
 =====================================================
-Every organization must have at least one environment. Every organization starts out with a single environment that is named ``_default``, which ensures that at least one environment is always available to the Chef Infra Server. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can onl ...
+Every Chef Infra Server organization must have at least one environment. Each organization starts out with a single environment that is named ``_default``. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can only be associated with a custom environment.
```
```suggestion
Every Chef Infra Server organization must have at least one environment. Each organization starts out with a single ``_default`` environment. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can only be associated with a custom environment.
```

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -5,21 +5,21 @@ About Environments
-Every organization must have at least one environment. Every organization starts out with a single environment that is named ``_default``, which ensures that at least one environment is always available to the Chef Infra Server. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can onl ...
+Every Chef Infra Server organization must have at least one environment. Each organization starts out with a single environment that is named ``_default``. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can only be associated with a custom environment.
 
 Environment Attributes
```
This section comes out of the blue. ELIF why we're suddenly discussing attributes.

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -97,33 +92,21 @@ Attribute precedence, when viewed as a table:
+Cookbook Pinning
+=====================================================
+
+Cookbook versions can be pinned per evironment. This allows for the controlled rollout of new cookbook releases through environments for testing before new versions reach production environments. See the environment format examples below for the cookbook pinning syntax.
```
```suggestion
Cookbook versions can be pinned in each environment, which allows you to control the rollout of new cookbook releases through successive testing environments before releasing new cookbook versions into production environments. See the environment format examples below for the cookbook pinning syntax.
```

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -97,33 +92,21 @@ Attribute precedence, when viewed as a table:
 =====================================================
-Environment data may be stored in two formats:
+
+Environments may be stored on disk (any in source control) in two formats:
```
```suggestion
Store environments on disk (in any source control) in one of two formats:
```

tas50 replied:
We said may because it's entirely optional to store them on disk. It's the right thing to do, but still optional.

kagarmoe replied:
I'll think about different words.

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -382,8 +364,37 @@ A node is considered to be associated with an environment when the ``chef_enviro
-Set using chef-solo
+Move Nodes
 -----------------------------------------------------
+Nodes can be moved between environments, such as from a "dev" to a "production" environment by using the ``knife exec`` subcommand. For example:
```
```suggestion
Use the ``knife exec`` subcommand to move nodes between environments, such as from a "dev" to a "production" environment. For example:
```

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -5,21 +5,21 @@ About Environments
 .. tag environment_attribute
 
-An attribute can be defined in an environment and then used to override the default settings on a node. When an environment is applied during a Chef Infra Client run, these attributes are compared to the attributes that are already present on the node. When the environment attributes take precedence over the default attributes, Chef Infra Client applies those new settings and values during a Chef ...
+Attributes can be defined in an environment and then used to override the default attributes in a cookbook. When an environment is applied during a Chef Infra Client run, environment attributes are compared to the attributes that are already present on the node. When the environment attributes take precedence over the default attributes, Chef Infra Client applies those new settings and values dur ...
```
Defining attributes in an environment lets you override the default attributes in a cookbook. When a Chef Infra Client run applies an environment, the client compares the environment attributes to the attributes that are already defined on the node [IN THE COOKBOOK? WHY ARE WE SUDDENLY TALKING ABOUT NODES?] Environment attributes set to ``override`` take precedence over the default attributes set on the node by cookbooks. Chef Infra Client applies the ``override`` environment attributes and values during a Chef Infra Client run.

...or something like this

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -5,21 +5,21 @@ About Environments
+Attributes can be defined in an environment and then used to override the default attributes in a cookbook. When an environment is applied during a Chef Infra Client run, environment attributes are compared to the attributes that are already present on the node. When the environment attributes take precedence over the default attributes, Chef Infra Client applies those new settings and values dur ...
 
-An environment attribute can only be set to be a default attribute or an override attribute. An environment attribute cannot be set to be a ``normal`` attribute. Use the ``default_attribute`` and ``override_attribute`` methods in the Ruby DSL file or the ``default_attributes`` and ``override_attributes`` hashes in a JSON data file.
+Environment attributes can be set to either ``default`` attribute level or an ``override`` attribute level.
```
Maybe an example would help?

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -53,16 +53,11 @@ There are two types of attributes that can be used with environments:
        .. end_tag
 
-Attribute Persistence
+Attribute Precedence
```
"Precedence" is a much better word! 👍

kagarmoe commented on chef_master/source/environments.rst:
```diff
@@ -53,16 +53,11 @@ There are two types of attributes that can be used with environments:
-All attributes except for normal attributes are reset at the beginning of a Chef Infra Client run. Chef Infra Client rebuilds these attributes using automatic attributes collected by Ohai at the beginning of each Chef Infra Client run, and then uses default and override attributes that are specified in cookbooks, roles, environments, and Policyfiles. All attributes are then merged and applied to  ...
 
-.. end_tag
+Environments are one of several locations where attributes can be applied in Chef Infra. It's important to understand how the precedence level works in order to understand, which attributes will be applied when Chef Infra Client runs.
```
Environments are one of several different places [FILES?] for defining attributes in Chef Infra. Attributes are defined in cookbook attribute files, recipes, environments, roles, and also in Ohai.  Attribute _precedence_ explains why an attribute defined in one file wins out over another during a Chef Infra Client run and gets applied to a node.

kagarmoe replied:
I think we should reverse the order of the attribute precedence list, so #1 is the Ohai automatic attribute.

kagarmoe reviewed (commented):
Mostly exploratory comments and revisions.

tas50 committed:
Update chef_master/source/environments.rst

Co-Authored-By: Kimberly Garmoe <kgarmoe@chef.io>

Signed-off-by: Tim Smith <tsmith@chef.io>
Sign-off-executed-by: Tim Smith <tsmith84@gmail.com>
Approved-at: behalf

tas50 committed:
Update chef_master/source/environments.rst

Co-Authored-By: Kimberly Garmoe <kgarmoe@chef.io>

Signed-off-by: Tim Smith <tsmith@chef.io>
Sign-off-executed-by: Tim Smith <tsmith84@gmail.com>
Approved-at: behalf

tas50 committed:
Update chef_master/source/environments.rst

Co-Authored-By: Kimberly Garmoe <kgarmoe@chef.io>

Signed-off-by: Tim Smith <tsmith@chef.io>
Sign-off-executed-by: Tim Smith <tsmith84@gmail.com>
Approved-at: behalf

tas50 committed:
Update the environment description to actually describe what it does

Move the _default junk elsewhere and focus on what an environment actually does for a user.

Signed-off-by: Tim Smith <tsmith@chef.io>

tas50 committed:
More description updates

Signed-off-by: Tim Smith <tsmith@chef.io>

Commit:
Merge branch 'master' into environments_v2
